[MIG][14.0] account_invoice_facturx

alexis-via opened the pull request:
This is not a "stupid" migration. Several additions in the XML of Factur-X:
- add 3rd address line if partner_address_street3 is installed
- add H.S. code and country of origin of product if product_harmonized_system is installed
and it also has several fixes:
- no error if there is no tax on an invoice line
- proper display of XML in the logs

alexis-via committed:
Rename account_invoice_zugferd to account_invoice_factur-x

Rename account_invoice_import_zugferd to account_invoice_import_factur-x
Add module to support py3o reporting engine:

alexis-via committed:
Fix field name on invoice

alexis-via committed:
Stop using precision_get('Account') (this decimal precision doesn't exist in v10) in account_invoice_factur-x and account_invoice_ubl

alexis-via committed:
Update test suite

alexis-via committed:
account_invoice_factur-x: Add support for Factur-X standard instead of ZUGFeRD

alexis-via committed:
Adapt test suite

alexis-via committed:
Add Factur-X_extension_schema.xmp for PDF metadata of Factur-X invoices

alexis-via committed:
account_invoice_factur-x: TestIndicator not in new XML

alexis-via committed:
Improve XML for Factur-X

Make country_id field required on res.partner form view

alexis-via committed:
Remove bank name tag, not allowed in Factur-X

alexis-via committed:
Add first support for the import of Factur-X invoices

Add support for start/end dates on lines
Better handling of invalid IBANs

alexis-via committed:
Add support for allowance and charges on Factur-X invoice import

To avoid headaches with invalid files, now check XSD of XML invoices before parsing it upon import
Move Factur-X/zugferd xsd to base_zugferd

alexis-via committed:
Add support for 'DueDateTypeCode' on Factur-X invoice generation

alexis-via committed:
account_invoice_factur-x: now generate refunds with document type = 381 instead of document type = 380 with negative amounts

Improve the code of refund import to support both negative invoices and positive refunds

alexis-via committed:
Start to use the new factur-x lib

alexis-via committed:
Improve + modularize generation of UBL (up-port from v8 PR)

Add script mass_invoice_import.py
Code cleanup

alexis-via committed:
Finalise the use of the FacturX lib

One of the zugferd-specific code has been standardised in UNECE, so moved to the account_tax_unece module

alexis-via committed:
Adapt to final version of factur-x lib

alexis-via committed:
Adapt to latest changes in the agreement_* modules

alexis-via committed:
account_invoice_factur-x: add XML tags for sepa direct debit

Rapid hack to make py3O work again

alexis-via committed:
FIX Factur-x invoice generation with Qweb

alexis-via committed:
account_invoice_factur-x use factur-x lib instead of PyPDF2

alexis-via committed:
Add logging in test to try to understand where it stalls on Travis

alexis-via committed:
account_invoice_factur-x: Replace full test that doesn't work on Travis because get_pdf() doesn't work on Travis (I don't know why) but a more simple test only on XML generation

alexis-via committed:
Update README of Factur-X modules

Commit:
restore get_pdf() in tests (#31)

* sale_order_ubl + purchase_order_ubl: restore get_pdf() in tests using HttpCase

* Restore get_pdf() in tests of account_invoice_factur-x and account_invoice_ubl modules

* Update oca_dependencies.txt

Commit:
account_invoice_factur-x* : use factur-x lib v0.2 (#32)

With this newer implementation, account_invoice_factur-x_py3o now really works

alexis-via committed:
Factur-X invoice generation: level is now configurable

Factur-X refund generation: type 380 or 381 is now a config option
Fix some mistakes in Factur-X XML
Update XSD to latest version

alexis-via committed:
FIX lang in XML translatable labels

Workaround bug in Factur-X specs or XSD
Factur-X generation fix

alexis-via committed:
FIX official Factur-X XSD (patch sent to FNFE-MPE for upstream correction)

Restore internal note at BasicWL level

alexis-via committed:
Add module account_e-invoice_generate to ensure compatibility for XML embedding in PDF between account_invoice_factur-x and account_invoice_ubl

alexis-via committed:
Add support for Factur-X logos to include in the visual part of Factur-X invoices

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

alexis-via committed:
acc_type on res.partner.bank is not stored any more, cf https://github.com/OCA/bank-payment/pull/467

alexis-via committed:
in v9+, partner.name is NOT a required field any more

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
Several fixes + import invoices with VAT in a company that cannot deduct VAT (#64)

* Move code to a new dedicated method
* FIX update of existing draft invoice
  Improve error message of partner matching
  Move hook of partner matching to put it after VAT matching, which is the
  most reliable
* Add support for importing an invoice with VAT in a company that cannot deduc VAT
  Display nice error when account is missing on product
* Improve compliancy with Factur-X standard (using schematron)

oca-travis committed:
Update account_invoice_factur-x.pot

alexis-via committed:
Remove 'CalculationPercent' and 'BasisAmount' which are only available for global charges and not for charges per line

oca-travis committed:
Update account_invoice_factur-x.pot

Commit:
Translated using Weblate (French)

Currently translated at 100,0% (26 of 26 strings)

Translation: edi-10.0/edi-10.0-account_invoice_factur-x
Translate-URL: https://translation.odoo-community.org/projects/edi-10-0/edi-10-0-account_invoice_factur-x/fr/

alexis-via committed:
Update for Factur-X specs 1.3

oca-transbot committed:
Update translation files

Updated by Actualizar ficheiros PO com o novo POT (msgmerge) hook in Weblate.

OCA-git-bot committed:
icon.png

OSevangelist committed:
Added translation using Weblate (German)

OSevangelist committed:
Translated using Weblate (German)

Currently translated at 100.0% (26 of 26 strings)

Translation: edi-10.0/edi-10.0-account_invoice_factur-x
Translate-URL: https://translation.odoo-community.org/projects/edi-10-0/edi-10-0-account_invoice_factur-x/de/

astirpe committed:
Temporarily disable tests of account_invoice_factur-x

alexis-via committed:
Port Factur-X invoice generation stack v10 -> v12

alexis-via committed:
Small fixes/improvements in Factur-X XML generation

alexis-via committed:
account_invoice_facturx: fix small error in XML generation

alexis-via committed:
Python3 fix in Factur-X XML generation

sbidoul committed:
[FIX] rst errors

OCA-git-bot committed:
[UPD] README.rst

oca-travis committed:
[UPD] Update account_invoice_facturx.pot

alexis-via committed:
[MIG] account_invoice_facturx to v13

MIG account_invoice_facturx_py3o to v13
Up-port improvements of account_e-invoice_generate from v12 to v13

gurneyalex committed:
run precommit hooks

extract a method from _cii_add_invoice_line_block to lower the
complexity rating of that method and get a green check.

alexis-via committed:
manifest: Use exact pypi package names

as suggested by Stéphane Bidoul

gurneyalex committed:
fix bad refactoring

alexis-via committed:
account_invoice_facturx/_py3o: black reformatting

OCA-git-bot committed:
[UPD] README.rst

alexis-via committed:
[MIG] account_invoice_facturx to v14

Several additions in the XML of Factur-X:
- add 3rd address line if partner_address_street3 is installed
- add H.S. code and country of origin of product if product_harmonized_system is installed

Several fixes:
- no error if there is no tax on an invoice line
- proper display of XML in the logs

thomaspaulb commented on account_invoice_facturx/models/account_move.py:
```diff
@@ -0,0 +1,901 @@
+                    tax, line_trade_settlement, ns
+                )
+        else:
+            self._cii_line_applicable_trade_tax_block(None, line_trade_settlement, ns)
```
Instead of adding an empty tax block, maybe consider doing what I did in https://github.com/OCA/edi/pull/282 when faced a similar issue for UBL generation; case in point is where the company is a VAT-exempt government institution which doesnt bother adding any taxes to invoices since they are exempt from it; but UBL requires the field also, so in that case it picks the default tax. The default tax they select has to be marked with a specific UNECE code that explains _why_ they are tax-exempt, so that code can then be neatly filled in the UNECE field on that default tax.

alexis-via committed:
account_invoice_facturx: add CountrySubDivisionName is partner block

alexis-via committed:
account_invoice_facturx: adapt to factur-x lib 2.x

Add incoterm in extended profile
Update variable names to make it consistant with py3

manengel commented:
Hello @alexis-via,
I saw that you worked on that PR two weeks ago. Is this one you still actively pursue? Do you have news for us?

alexis-via commented:
This PR works well, you can use it in production on odoo v14 (I do).
I plan to implement order-X, so a lot of code of this module will be moved from account_invoice_facturx to base_facturx, to be used by the order-X module. But it can be done in a later PR.

alexis-via commented:
Travis should be green this time...let's see.

qpeyramale commented:
LGTM

manengel commented:
What is still missing in order to finalize this PR? Is it already foreseeable this will get merged into the 14.0 branch?

manengel commented:
@alexis-via Is there a way somebody could help you in order to finalize this PR?

fshah-initos commented:
Hi @alexis-via 
can you please resolve the conflicts?

fshah-initos commented:
Hi @alexis-via
can you please resolve the conflicts?

hkapatel-initos commented:
Hi @alexis-via
can you please make Travis happy?

hkapatel-initos commented:
Hi @alexis-via
can you please make Travis happy?

hkapatel-initos commented:
Hi @alexis-via
can you please make Travis happy?

alexis-via committed:
account_invoice_facturx: add prepare method for attachments

dsolanki-initos commented:
Hi @alexis-via
it would be nice if all checks are happy?

dsolanki-initos commented:
Hi @alexis-via can you make Travis happy?

nibraun reviewed (approved):
Top

alexis-via commented:
If someone has an idea why the tests get stalled on Travis, please tell us ! The tests run fine on my laptop.

dsolanki-initos commented:
Hi @alexis-via can you make Travis happy?

alexis-via commented:
@dsolanki-initos As you can see in my last message, I'm seeking for some technical help to understand why the tests get stalled on Travis. Please stop repeating "can you xxx" every few days.

simahawk commented:
```
2021-10-30 06:21:25,310 6652 CRITICAL openerp_test odoo.service.server: Failed to initialize database `openerp_test`.

Traceback (most recent call last):

File "/home/travis/odoo-14.0/odoo/service/server.py", line 1199, in preload_registries

registry = Registry.new(dbname, update_module=update_module)

File "/home/travis/odoo-14.0/odoo/modules/registry.py", line 89, in new

odoo.modules.load_modules(registry._db, force_demo, status, update_module)

File "/home/travis/odoo-14.0/odoo/modules/loading.py", line 426, in load_modules

modules.button_install()

File "<decorator-gen-70>", line 2, in button_install

File "/home/travis/odoo-14.0/odoo/addons/base/models/ir_module.py", line 74, in check_and_log

return method(self, *args, **kwargs)

File "/home/travis/odoo-14.0/odoo/addons/base/models/ir_module.py", line 431, in button_install

raise UserError(msg % (module.shortdesc, exclusion.exclusion_id.shortdesc))

odoo.exceptions.UserError: Modules "Account Invoice Import Simple PDF" and "Account Invoice Import Invoice2data" are incompatible.

2021-10-30 06:21:25,311 6652 INFO openerp_test odoo.service.server: Initiating shutdown

2021-10-30 06:21:25,311 6652 INFO openerp_test odoo.service.server: Hit CTRL-C again or send a second signal to force the shutdown.

----------

Failed to load registry

Failed to initialize database `openerp_test`.

Traceback (most recent call last):

  File "/home/travis/odoo-14.0/odoo/service/server.py", line 1199, in preload_registries

    registry = Registry.new(dbname, update_module=update_module)

  File "/home/travis/odoo-14.0/odoo/modules/registry.py", line 89, in new

    odoo.modules.load_modules(registry._db, force_demo, status, update_module)

  File "/home/travis/odoo-14.0/odoo/modules/loading.py", line 426, in load_modules

    modules.button_install()

  File "<decorator-gen-70>", line 2, in button_install

  File "/home/travis/odoo-14.0/odoo/addons/base/models/ir_module.py", line 74, in check_and_log

    return method(self, *args, **kwargs)

  File "/home/travis/odoo-14.0/odoo/addons/base/models/ir_module.py", line 431, in button_install

    raise UserError(msg % (module.shortdesc, exclusion.exclusion_id.shortdesc))

odoo.exceptions.UserError: Modules "Account Invoice Import Simple PDF" and "Account Invoice Import Invoice2data" are incompatible.

Message not found: 'Modules loaded.'
```
I ran the last job again and I got this.

alexis-via commented:
@simahawk Strange, this error should not happen because account_invoice_import_invoice2data is excluded in the travis config file.
The branch origin/14.0 is currently green. I rebased and the tests now stop on:
```
2021-10-31 08:40:06,697 6638 INFO openerp_test odoo.modules.loading: Loading module account_invoice_facturx (90/96)

2021-10-31 08:40:07,046 6638 INFO openerp_test odoo.modules.registry: module account_invoice_facturx: creating or updating database tables

2021-10-31 08:40:07,460 6638 INFO openerp_test odoo.modules.loading: loading account_invoice_facturx/views/res_partner.xml

2021-10-31 08:40:07,598 6638 INFO openerp_test odoo.modules.loading: loading account_invoice_facturx/views/res_config_settings.xml

2021-10-31 08:40:07,728 6638 INFO openerp_test odoo.modules.loading: loading account_invoice_facturx/views/report_invoice.xml

2021-10-31 08:40:07,753 6638 INFO openerp_test odoo.modules.loading: Module account_invoice_facturx: loading demo

2021-10-31 08:40:07,768 6638 INFO openerp_test odoo.addons.account_invoice_facturx.tests.test_facturx_invoice: Starting TestFacturXInvoice.test_deep_customer_invoice ...

2021-10-31 08:40:09,859 6638 INFO ? werkzeug: 127.0.0.1 - - [31/Oct/2021 08:40:09] "GET /account_invoice_facturx/static/logos/factur-x-minimum.png HTTP/1.1" 200 - - - -
```

There is no error, it just stops there.

simahawk commented:
talking about logs, careful with that. They slow down the build for huge content and they might prevent tests to run till the end.
If you are dealing w/ binary files here, make sure you don't log them.
Example: https://github.com/OCA/edi/pull/468

alexis-via commented:
@simahawk it was just temporary logs to spot where the tests were blocked. I figured out that it was blocked in _render_qweb_pdf(). So I changed the test to only test XML generation. It's better than nothing. 
If someone knows how to make _render_qweb_pdf() work with Travis, he can step in and propose a new PR. But I received more "can you please make Travis green" than "Here is a patch to make Travis green"...

alexis-via commented:
/ocabot merge nobump
